File: 1655.cpp
/* 시간 초과 실패 */
#include <iostream>

using namespace std;

typedef struct {
    int* heapq;
    int index;
} Heap;

void heappush(Heap* heap, int item) {
    heap->heapq[heap->index] = item;
    int parent = heap->index / 2;;

    while (parent != 0 && heap->heapq[parent] < item) {
        int tmp = heap->heapq[parent];
        heap->heapq[parent] = item;
        heap->heapq[heap->index] = tmp;

        parent /= 2;
    }

    heap->index++;
}

int heappop(Heap* heap) {
    int i = --(heap->index);
    int ret = heap->heapq[1];
    heap->heapq[1] = heap->heapq[i];

    int child = 2;
    int parent = 1;
    int tmp;
    while(child < i){
        if(child+1 < i && heap->heapq[child] < heap->heapq[child+1]) {
            child++;
        }

        if (heap->heapq[child] > heap->heapq[parent]) {
            tmp = heap->heapq[child];
            heap->heapq[child] = heap->heapq[parent];
            heap->heapq[parent] = tmp;

            parent = child;
            child = parent*2;
        }
    }

    return ret;
}

int main(){
    int N;
    int mid = 10001;

    scanf("%d", &N);

    Heap minheap;
    minheap.heapq = new int[N/2 +2];
    minheap.index = 1;
    Heap maxheap;
    maxheap.heapq = new int[N/2 +2];
    maxheap.index = 1;

    for (int i = 0; i < N; i++) {
        int tmp;
        scanf("%d", &tmp);

        if (tmp > mid) {
            heappush(&minheap, -tmp);

            if (minheap.index > maxheap.index) {
                tmp = heappop(&minheap);
                heappush(&maxheap, -tmp);
            }
        }
        else {
            heappush(&maxheap, tmp);

            if (maxheap.index-1 > minheap.index) {
                tmp = heappop(&maxheap);
                heappush(&minheap, -tmp);
            }
        }

        mid = maxheap.heapq[1];
        printf("%d\n", mid);
    }
}

File: 1655.py
import heapq
import sys
input = sys.stdin.readline

N = int(input())

minheap = []
maxheap = []
mid = 10001

lmin = 0
lmax = 0

for i in range(N):
    tmp = int(input())

    if tmp > mid:
        heapq.heappush(minheap, tmp)
        lmin += 1
        
        if lmin > lmax:
            pop = heapq.heappop(minheap)
            heapq.heappush(maxheap, -pop)
            lmin -= 1
            lmax += 1
            
    else:
        heapq.heappush(maxheap, -tmp)
        lmax += 1
        
        if lmax - 1 > lmin:
            pop = heapq.heappop(maxheap)
            heapq.heappush(minheap, -pop)
            lmax -= 1
            lmin += 1
            
    mid = -maxheap[0]
    print(mid)